Mathlib/Computability/GNFA.lean

```diff
@@ -0,0 +1,383 @@
+/-
+Copyright (c) 2022 Russell Emerine, 2024 Tom Kranz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Russell Emerine, Tom Kranz
+-/
+import Mathlib.Computability.RegularExpressions
+import Mathlib.Data.FinEnum.Option
+
+#align_import computability.GNFA
+
+/-!
+# Generalized Nondeterministic Finite Automata
+
+This file contains the definition of a Generalized Nondeterministic Finite Automaton, a state
+machine which determines whether a string (implemented as a `List` over an arbitrary alphabet) is in
+a regular set by evaluating the string over its net of regular expressions. We show
+that GNFA's are equivalent to `NFA`'s, and that GNFA's are equivalent to smaller GNFA's with a state
+"ripped" out. Through this mechanism, we show that `NFA`'s are equivalent to `RegularExpression`s.
+
+## References
+
+TODO: someone please tell me how to best cite this file?
+* <https://courses.engr.illinois.edu/cs373/sp2013/Lectures/lec07.pdf>
```

Comment:
I guess an entry in the `references.bib` is reasonable; there are probably a few other url-based references there you could copy.

Reply:
This is a remnant of Russell's. I've been wondering about that source since I first saw it, since it conveys pretty basic knowledge, not even bestowing a name upon the RE-NFA trafo, and doesn't seem to inspire any of the ingenuity of Russell's actual implementation or proof. They don't seem to have updated the docstrings from the first draft (there was something about only permitting `Fin`s as state types) and this was probably the initial lecture where they learned about the constructions. I didn't want to just replace it without prior discussion, but I would suggest just replacing it with a textbook reference. The lecture seems to be based on Sipser's Intro to the Theory of Computation, down to the two failed attempts at constructing a Kleene-starred NFA, but it's of course impossible to tell.